torch/fx/_graph_pickler.py

```diff
@@ -0,0 +1,568 @@
+import dataclasses
+import importlib
+import io
+import pickle
+from abc import abstractmethod
+from typing import Any, Callable, Dict, NewType, Optional, Tuple, Type, TypeVar, Union
+from typing_extensions import override, Self
+
+import torch
+import torch.utils._pytree as pytree
+from torch._guards import TracingContext
+from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode, Tensor
+from torch._subclasses.meta_utils import (
+    MetaConverter,
+    MetaTensorDesc,
+    MetaTensorDescriber,
+)
+from torch.fx.experimental.sym_node import SymNode
+from torch.fx.experimental.symbolic_shapes import ShapeEnv
+from torch.utils._mode_utils import no_dispatch
+
+
+_SymNodeT = TypeVar("_SymNodeT", torch.SymInt, torch.SymFloat)
+
+
+class GraphPickler(pickle.Pickler):
+    """
+    GraphPickler is a Pickler which helps pickling fx graph - in particular
+    GraphModule.
+    """
+
+    def __init__(self, file: io.BytesIO) -> None:
+        super().__init__(file, protocol=pickle.HIGHEST_PROTOCOL)
+
+        # This abomination is so we can pass external decoding state to the
+        # unpickler functions. We serialize _unpickle_state as a persistent
+        # external item and when we deserialize it we return the common state
+        # object.
+        self._unpickle_state = _UnpickleStateToken(object())
+
+        # This is used to describe tensors. It needs to be common across the
+        # pickle so that duplicates and views are properly handled.
+        self._meta_tensor_describer = MetaTensorDescriber(copy_data=False)
+
+    @override
+    def reducer_override(
+        self, obj: object
+    ) -> Tuple[Callable[..., Any], Tuple[Any, ...]]:
+        # This function is supposed to return either NotImplemented (meaning to
+        # do the default pickle behavior) or a pair of (unpickle callable, data
+        # to pass to unpickle).
+
+        # We could instead teach individual classes how to pickle themselves but
+        # that has a few problems:
+        #
+        #   1. If we have some special needs (maybe for this use-case we don't
+        #      want to fully serialize every field) then we're adding private
+        #      details to a public interface.
+        #
+        #   2. If we need to have some common shared data (such as a
+        #      FakeTensorMode) which is passed to each value it's harder to
+        #      support.
+
+        # These are the types that need special handling. See the individual
+        # *PickleData classes for details on pickling that particular type.
```

Comment:
- What's the recursive behavior of this interface? Let's say I pickle a FakeTensor and it goes to TensorPickleData.reduce_helper. In the meta of that FakeTensor there's a sym int or shape env. Does it call back into this reducer_override? Similar question if something goes wrong and we fallback to default pickling: let's say we find a class Foo that we don't know how to pickle, and we fall back tot ehd efault pickler. Foo then contains another FakeTensor inside of it. Does it call reducer_override again? Or is it permanently default after that? 


- What about real tensors? Sometimes a graph module will have a real tensor embedded in it as a constant. Do we want a pickler for that?

Reply:
Yes - the reducer_override is recursively applied on the output.
Real tensors - maybe? The problem with real tensors is that it's not obvious we can unpickle them (and we'd need to pickle all the data which could be huge)

Reply:
Hm there's got to be a unit test that exhibits adding a real tensor as a constant in the graph module... is the behavior today that we would throw NotImplemented and use the default pickler?

Reply:
Correct - for any type that isn't explicitly handled here we *return* NotImplemented (not thrown!) which uses the default behavior.